src/test/java/com/zendesk/maxwell/MaxwellTestSupport.java

```diff
@@ -232,30 +234,46 @@ public void run() {
 
 		for ( ;; ) {
 			RowMap row = maxwell.poll(pollTime);
-			pollTime = 500L; // after the first row is receive, we go into a tight loop.
+			pollTime = 500L; // after the first row is received, we go into a tight loop.
 
-			if ( row == null ) {
-				LOGGER.debug("timed out waiting for final row.  Last position we saw: " + lastPositionRead);
-				break;
+			if ( row != null ) {
+				if ( row.toJSON(config.outputConfig) != null ) {
+					LOGGER.debug("getRowsWithReplicator: saw: " + row.toJSON(config.outputConfig));
+					list.add(row);
+				}
+				lastPositionRead = row.getPosition();
 			}
 
-			lastPositionRead = row.getPosition();
+			boolean replicationComplete = lastPositionRead != null && lastPositionRead.getLastHeartbeatRead() >= finalHeartbeat;
+			boolean bootstrapComplete = !bootstrapController.hasIncompleteTasks();
+			boolean timedOut = !replicationComplete && row == null;
 
-			if ( lastPositionRead != null && lastPositionRead.getLastHeartbeatRead() >= finalHeartbeat ) {
-				// consume whatever's left over in the buffer.
-				for ( ;; ) {
-					RowMap r = maxwell.poll(100);
-					if ( r == null )
-						break;
+			if (timedOut) {
+				LOGGER.debug("timed out waiting for final row. Last position we saw: " + lastPositionRead);
+				break;
+			}
 
-					if ( r.toJSON(config.outputConfig) != null )
-						list.add(r);
+			if (bootstrapComplete && replicationComplete) {
+				// For sanity testing, we wait another 2s to verify that no additional changes were pending.
+				// This slows down the test suite, so only enable when debugging.
+				boolean checkHasNoPendingChanges = false;
+				if (checkHasNoPendingChanges) {
+					long deadline = System.currentTimeMillis() + 2000;
+					while(true) {
```

Comment:
could this just be `while(System.currentTimeMillis() <= deadline)`?  or a `do /while`?